lib/mpl_toolkits/mplot3d/art3d.py

```diff
@@ -1086,3 +1115,84 @@ def _zalpha(colors, zs):
     sats = 1 - norm(zs) * 0.7
     rgba = np.broadcast_to(mcolors.to_rgba_array(colors), (len(zs), 4))
     return np.column_stack([rgba[:, :3], rgba[:, 3] * sats])
+
+
+def _generate_normals(polygons):
+    """
+    Compute the normals of a list of polygons.
+
+    Normals point towards the viewer for a face with its vertices in
+    counterclockwise order, following the right hand rule.
+
+    Uses three points equally spaced around the polygon.
+    This normal of course might not make sense for polygons with more than
+    three points not lying in a plane, but it's a plausible and fast
+    approximation.
+
+    Parameters
+    ----------
+    polygons : list of (M_i, 3) array-like, or (..., M, 3) array-like
+        A sequence of polygons to compute normals for, which can have
+        varying numbers of vertices. If the polygons all have the same
+        number of vertices and array is passed, then the operation will
+        be vectorized.
+
+    Returns
+    -------
+    normals : (..., 3) array
+        A normal vector estimated for the polygon.
+    """
+    if isinstance(polygons, np.ndarray):
+        # optimization: polygons all have the same number of points, so can
```

Comment:
Good question. I must say that I do not fully following this. In the example, I originally didn't use an `np.array` and then it errored in the else-clause. However, that is still covered in the tests.

It may be that this needs additional work as we now open it up a bit, but I think we will need to collect bug reports of non-working cases and take it from there. (Note that the code is exactly the same as before, just that it is executed in a single location.)

Reply:
I *think* that the idea is that an `np.ndarray` is always the same size by construction?

```
An array object represents a multidimensional, homogeneous array of fixed-size items.
```